[GStreamer] Video frame converter uses the wrong gl context

carlosgcampos opened the pull request:
#### ba826db2307bf3f96accb418889ee5bcf83aac78
<pre>
[GStreamer] Video frame converter uses the wrong gl context
<a href="https://bugs.webkit.org/show_bug.cgi?id=287241">https://bugs.webkit.org/show_bug.cgi?id=287241</a>

Reviewed by Philippe Normand.

For some reason the gl context set on the sink element is not propagated
to others, so set it in the pipeline instead. The converter also fails
if the decoder is producing gl memory ax external EOS texture, because
the gldownload element can&apos;t handle that. So, instead of downloaing and
then converting, we now convert using glcolorconvert and then download,
since glcolorconvert will produce a normal texture. But glcolorconvert
can&apos;t handle system memory, so we now use a different pipeline depending
on the input caps, using previous pipeline for system memory. This patch
also removes the sample download in videoFrameForCurrentTime() and
GStreamerInternalVideoEncoder::encode since that&apos;s already handled by
the converter later when needed. The pipeline is now also released when
not used after 30 seconds.

* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt:
* LayoutTests/platform/wpe/media/media-source/media-source-paint-to-canvas-expected.txt: Removed.
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
(WebCore::GStreamerVideoFrameConverter::Pipeline::Pipeline):
(WebCore::GStreamerVideoFrameConverter::Pipeline::run):
(WebCore::GStreamerVideoFrameConverter::singleton):
(WebCore::GStreamerVideoFrameConverter::GStreamerVideoFrameConverter):
(WebCore::GStreamerVideoFrameConverter::ensurePipeline):
(WebCore::GStreamerVideoFrameConverter::convert):
(WebCore::GStreamerVideoFrameConverter::releaseUnusedSystemMemoryPipelineTimerFired):
(WebCore::GStreamerVideoFrameConverter::releaseUnusedGLMemoryPipelineTimerFired):
(WebCore::GStreamerVideoFrameConverter::releaseUnusedDMABufMemoryPipelineTimerFired):
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::videoFrameForCurrentTime):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::encode):

Canonical link: <a href="https://commits.webkit.org/290266@main">https://commits.webkit.org/290266@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2815ed7e52fca3585d41ba93e7340d4779a6929b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/89301 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/8825 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/44137 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/94287 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/40062 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/91352 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/9213 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/17147 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/68806 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/26474 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/92303 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/7070 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/81048 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/49167 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/6816 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/35433 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/39169 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/77172 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/36418 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/96116 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/16481 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/12165 "Found 2 new test failures: imported/w3c/web-platform-tests/css/selectors/invalidation/modal-pseudo-class-in-has.html imported/w3c/web-platform-tests/fullscreen/api/element-request-fullscreen-namespaces.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/77680 "Found 3 new test failures: imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource.html media/media-video-fullrange-offscreen.html media/media-video-fullrange.html (failure)") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/16737 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/76836 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/76979 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/19029 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/21396 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/19996 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/9616 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/16495 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/21806 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/16236 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/19687 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/18017 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
```diff
@@ -46,14 +46,14 @@ GStreamerVideoFrameConverter::GStreamerVideoFrameConverter()
-    gst_bin_add_many(GST_BIN_CAST(m_pipeline.get()), m_src.get(), gldownload, videoconvert, videoscale, m_sink.get(), nullptr);
-    gst_element_link_many(m_src.get(), gldownload, videoconvert, videoscale, m_sink.get(), nullptr);
+    gst_bin_add_many(GST_BIN_CAST(m_pipeline.get()), m_src.get(), glcolorconvert, gldownload, videoscale, m_sink.get(), nullptr);
+    gst_element_link_many(m_src.get(), glcolorconvert, gldownload, videoscale, m_sink.get(), nullptr);
```
This breaks webcodec tests, with glcolorconvert we no longer accept video/x-raw (gldownload does).

ERROR   webkitvideoframeconverter GStreamerVideoFrameConverter.cpp:97:convert:<video-frame-converter> Unable to convert video frame. Error: error message: 0x7fed34004b90, time 99:99:99.999999999, seq-num 227, element 'appsrc0', GstMessageError, gerror=(GError)NULL, debug=(string)"../libs/gst/base/gstbasesrc.c\(3177\):\ gst_base_src_loop\ \(\):\ /GstPipeline:video-frame-converter/GstAppSrc:appsrc0:\012streaming\ stopped\,\ reason\ not-negotiated\ \(-4\)", details=(structure)"details\,\ flow-return\=\(int\)-4\;";

philn replied:
That's triggered from `GStreamerInternalVideoEncoder::encode()`... I don't remember in which situation we have GL there (hence the downloadSample call)...

philn commented on Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
```diff
@@ -124,8 +166,14 @@ GRefPtr<GstSample> GStreamerVideoFrameConverter::convert(const GRefPtr<GstSample
 
+void GStreamerVideoFrameConverter::releaseUnusedPipelinesTimerFired()
+{
+    m_systemMemoryPipeline = nullptr;
```
Is this really sufficient?

Usually to cleanup a pipeline we set its state to NULL first, but maybe here it's not needed since `convert()` does it already.

When re-creating the pipeline though, I'm not sure what happens with m_src and m_sink... are they re-parented?

carlosgcampos replied:
I had the same doubts, so I tried with a shorter cleanup time and the pipeline was correctly destroyed/created. In any case I'm not yet convinced that a singleton is the best approach... But I want a simple patch for the stable branch.

carlosgcampos committed:
[GStreamer] Video frame converter uses the wrong gl context
https://bugs.webkit.org/show_bug.cgi?id=287241

Reviewed by Philippe Normand.

For some reason the gl context set on the sink element is not propagated
to others, so set it in the pipeline instead. The converter also fails
if the decoder is producing gl memory ax external EOS texture, because
the gldownload element can't handle that. So, instead of downloaing and
then converting, we now convert using glcolorconvert and then download,
since glcolorconvert will produce a normal texture. But glcolorconvert
can't handle system memory, so we now use a different pipeline depending
on the input caps, using previous pipeline for system memory. This patch
also removes the sample download in videoFrameForCurrentTime() and
GStreamerInternalVideoEncoder::encode since that's already handled by
the converter later when needed. The pipeline is now also released when
not used after 30 seconds.

* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/webcodecs/videoFrame-canvasImageSource-expected.txt:
* LayoutTests/platform/wpe/media/media-source/media-source-paint-to-canvas-expected.txt: Removed.
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.cpp:
(WebCore::GStreamerVideoFrameConverter::Pipeline::Pipeline):
(WebCore::GStreamerVideoFrameConverter::Pipeline::run):
(WebCore::GStreamerVideoFrameConverter::singleton):
(WebCore::GStreamerVideoFrameConverter::GStreamerVideoFrameConverter):
(WebCore::GStreamerVideoFrameConverter::ensurePipeline):
(WebCore::GStreamerVideoFrameConverter::convert):
(WebCore::GStreamerVideoFrameConverter::releaseUnusedSystemMemoryPipelineTimerFired):
(WebCore::GStreamerVideoFrameConverter::releaseUnusedGLMemoryPipelineTimerFired):
(WebCore::GStreamerVideoFrameConverter::releaseUnusedDMABufMemoryPipelineTimerFired):
* Source/WebCore/platform/graphics/gstreamer/GStreamerVideoFrameConverter.h:
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::videoFrameForCurrentTime):
* Source/WebCore/platform/graphics/gstreamer/VideoEncoderGStreamer.cpp:
(WebCore::GStreamerInternalVideoEncoder::encode):

Canonical link: https://commits.webkit.org/290266@main
